src/Microsoft.Health.TaskManagement/JobHosting.cs

```diff
@@ -0,0 +1,250 @@
+                    }
+                    catch (Exception completeEx)
+                    {
+                        _logger.LogError(completeEx, "Job {jobId} failed to complete.", jobInfo.Id);
+                    }
+
+                    return;
+                }
+                catch (Exception ex)
+                {
+                    _logger.LogError(ex, "Job {jobId} failed.", jobInfo.Id);
+
+                    object error = new { message = ex.Message };
+                    jobInfo.Result = JsonConvert.SerializeObject(error);
+                    jobInfo.Status = JobStatus.Failed;
+
+                    try
+                    {
+                        await _queueClient.CompleteJobAsync(jobInfo, false, CancellationToken.None);
+                    }
+                    catch (Exception completeEx)
+                    {
+                        _logger.LogError(completeEx, "Job {jobId} failed to complete.", jobInfo.Id);
+                    }
+
+                    return;
+                }
+
+                try
+                {
+                    jobInfo.Status = JobStatus.Completed;
+                    await _queueClient.CompleteJobAsync(jobInfo, true, CancellationToken.None);
+                    _logger.LogInformation("Job {jobId} completed.", jobInfo.Id);
+                }
+                catch (Exception completeEx)
+                {
+                    _logger.LogError(completeEx, "Job {jobId} failed to complete.", jobInfo.Id);
+                }
+            }
+            finally
+            {
+                _activeJobsNeedKeepAlive.Remove(jobInfo.Id, out _);
+            }
+        }
+
+        private async Task KeepAliveSingleJobAsync(JobInfo jobInfo, CancellationTokenSource jobCancellationToken)
+        {
+            try
+            {
+                bool shouldCancel = false;
+                try
+                {
+                    JobInfo keepAliveResult = await _queueClient.KeepAliveJobAsync(jobInfo, CancellationToken.None);
+                    shouldCancel |= keepAliveResult.CancelRequested;
+                }
+                catch (JobNotExistException notExistEx)
+                {
+                    _logger.LogError(notExistEx, "Job {jobId} not exist or runid not match.", jobInfo.Id);
+                    shouldCancel = true;
+                }
+
+                if (shouldCancel && !jobCancellationToken.IsCancellationRequested)
+                {
+                    jobCancellationToken.Cancel();
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError(ex, "Failed to keep alive on job {jobId}", jobInfo.Id);
+            }
+        }
+
+        private async Task KeepAliveJobsAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Start to keep alive job message.");
```

Comment:
>message." [](http://example.com/codeflow?start=60&length=9)

We should include job id

Reply:
this log is only for the keep alive thread start it is not for specific job. and looks like too many logs for every keep alive call?